engine/src/main/java/org/pentaho/di/cluster/SlaveServer.java

```diff
@@ -0,0 +1,1403 @@
+      return;
+    }
+    HttpHost httpHost = new HttpHost( proxyHost, Integer.valueOf( proxyPort ) );
+
+    RequestConfig requestConfig = RequestConfig.custom()
+      .setProxy( httpHost )
+      .build();
+
+    context.setRequestConfig( requestConfig );
+  }
+
+  /**
+   *
+   * @return HttpClientContext with authorization credentials
+   */
+  protected HttpClientContext getAuthContext() {
+    HttpClientContext context = HttpClientContext.create();
+    addCredentials( context );
+    addProxy( context );
+    return context;
+  }
+
+  /**
+   * @return the master
+   */
+  public boolean isMaster() {
+    lock.readLock().lock();
+    try {
+      return master;
+    } finally {
+      lock.readLock().unlock();
+    }
+  }
+
+  /**
+   * @param master the master to set
+   */
+  public void setMaster( boolean master ) {
+    lock.writeLock().lock();
+    this.master = master;
+    lock.writeLock().unlock();
+
+  }
+
+  public String execService( String service, boolean retry ) throws Exception {
+    int tries = 0;
+    int maxRetries = 0;
+    if ( retry ) {
+      maxRetries = KETTLE_CARTE_RETRIES;
+    }
+    while ( true ) {
+      try {
+        return execService( service );
+      } catch ( Exception e ) {
+        if ( tries >= maxRetries ) {
+          throw e;
+        } else {
+          try {
+            Thread.sleep( getDelay( tries ) );
+          } catch ( InterruptedException e2 ) {
+            //ignore
+          }
+        }
+      }
+      tries++;
+    }
+  }
+
+  public static long getDelay( int trial ) {
+    long current = KETTLE_CARTE_RETRY_BACKOFF_INCREMENTS;
+    long previous = 0;
+    for ( int i = 0; i < trial; i++ ) {
+      long tmp = current;
+      current = current + previous;
+      previous = tmp;
+    }
+    return current + RANDOM.nextInt( (int) Math.min( Integer.MAX_VALUE, current / 4L ) );
+  }
+
+  public String execService( String service ) throws Exception {
+    return execService( service, new HashMap<>() );
+  }
+
+  // Method is defined as package-protected in order to be accessible by unit tests
+  String getResponseBodyAsString( InputStream is ) throws IOException {
+    BufferedReader bufferedReader = new BufferedReader( new InputStreamReader( is ) );
+    StringBuilder bodyBuffer = new StringBuilder();
+    String line;
+
+    try {
+      while ( ( line = bufferedReader.readLine() ) != null ) {
+        bodyBuffer.append( line );
+      }
+    } finally {
+      bufferedReader.close();
+    }
+
+    return bodyBuffer.toString();
+  }
+
+  // Method is defined as package-protected in order to be accessible by unit tests
+  HttpGet buildExecuteServiceMethod( String service, Map<String, String> headerValues )
+    throws UnsupportedEncodingException {
+    HttpGet method = new HttpGet( constructUrl( service ) );
+
+    for ( String key : headerValues.keySet() ) {
+      method.setHeader( key, headerValues.get( key ) );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.